lightly/models/swav.py

```diff
@@ -0,0 +1,55 @@
+import torch
+from torch import nn
+
+from lightly.models.modules import SwaVProjectionHead, SwaVPrototypes
+from lightly.loss.memory_bank import MemoryBankModule
+
+
+class SwaV(nn.Module):
+    def __init__(self, backbone, num_ftrs, out_dim, n_prototypes, queue_length=0):
+        super().__init__()
+        self.backbone = backbone
+        self.projection_head = SwaVProjectionHead(num_ftrs, num_ftrs, out_dim)
+        self.prototypes = SwaVPrototypes(out_dim, n_prototypes=n_prototypes)
+        # Queues are initialized in the first forward call
+        self.queues = []
+        self.queue_length = queue_length
+
+    def forward(self, high_resolution, low_resolution):
+        self._create_queues(n_queues=len(high_resolution), device=high_resolution[0].device)
+
+        self.prototypes.normalize()
+
+        high_resolution_features = [self._subforward(x) for x in high_resolution]
+        low_resolution_features = [self._subforward(x) for x in low_resolution]
+        queue_features = self._get_queue_features(high_resolution_features)
+
+        high_resolution_prototypes = [self.prototypes(x) for x in high_resolution_features]
+        low_resolution_prototypes = [self.prototypes(x) for x in low_resolution_features]
+        queue_prototypes = [self.prototypes(x) for x in queue_features]
+
```

Comment:
I believe this line should be in a `torch.no_grad()` block:
```
queue_prototypes = [self.prototypes(x) for x in queue_features]
```

The original repo does queue prototype assignment also with `torch.no_grad()`: https://github.com/facebookresearch/swav/blob/06b1b7cbaf6ba2a792300d79c7299db98b93b7f9/main_swav.py#L291-L301